### ISSUE TITLE
Collect & visualise spark-stats API usage based on access tokens

### ISSUE THREAD

**Issue by bajtos** [space-meridian/roadmap#206]
We started to hand out API keys to people who signed up with their email for Spark Data offering. Now we would like to understand which user is using Spark API and how frequently.

Docs which the users see: https://filspark.com/api-docs

Notes:

- We are not enforcing the access token yet, anonymous usage is allowed. We want to track anonymous requests too.
- The API is behind Cloudflare cache, some requests are served from the cache and don't hit our spark-stats service. We want to include the cached requests in the data.
- Our API docs used to be slightly wrong. We need to look for the API key in the `api-key` header & `api-key` query string in addition to the standard `Authorization` header in `Bearer` format.

### TODO:
- [x] Initial implementation
- [x] #211
- [x] Dashboard showing the collected data

Update our internal spark-stats consumers to send an access token:
- [x] https://spacemeridian.grafana.net/
- [ ] https://grafana.filstation.app/
- [x] https://dashboard.filspark.com/
- [ ] what else?

**Comment by pyropy** [comment]
Are we currently using any tools for monitoring and collecting logs? Also, what service is responsible for managing API keys and setting rate limits?

**Comment by juliangruber** [comment]
Logs are being collected in Papertrail, which I believe Miro gave you access to.

I believe at the moment API keys are only parsed in Cloudflare, @bajtos correct me if I'm wrong.

**Comment by pyropy** [comment]
> Logs are being collected in Papertrail, which I believe Miro gave you access to.
> 
> I believe at the moment API keys are only parsed in Cloudflare, @bajtos correct me if I'm wrong.

I don't have access to Papertrail yet.

**Comment by pyropy** [comment]
I believe if we want to track both cached and non-cached requests we'd have to use  cloudflare workers. In that case we'd have two options:
- Push logs directly from cloudflare worker to papertrail
- Log on cloudflare worker and setup [cloudflare logpush](https://developers.cloudflare.com/workers/observability/logs/logpush/) that would forward cloudflare worker logs over to papertrail

**Comment by juliangruber** [comment]
Instead of pushing logs to papertrail, where we then would parse them (I assume), can we make the cloudflare worker submit information about the request into a DB, via an HTTP API that we create? It could also feed this data into InfluxDB, for time series processing

**Comment by pyropy** [comment]
> Instead of pushing logs to papertrail, where we then would parse them (I assume), can we make the cloudflare worker submit information about the request into a DB, via an HTTP API that we create? It could also feed this data into InfluxDB, for time series processing

Yeah I would prefer something like that over the papertrail also. It will be much easier to aggregate and visualize the data.

**Comment by juliangruber** [comment]
Cool, let's either add a route to `spark-api` then which persists this information in Postgres, or write it to InfluxDB. If both are viable options to you, best create a small document or GitHub comment describing your decision making process (pros cons, etc)

**Comment by bajtos** [comment]
Can we put the new route to `spark-stats` please? We are measuring spark-stats API usage, nothing from `spark-api` is involved in this.

**Comment by bajtos** [comment]
Writing from Cloudflare worker directly to InfluxDB may be the simplest option, as we don't need to write any new REST API 💪🏻

**Comment by juliangruber** [comment]
Good point, yes should be in `spark-stats`

**Comment by juliangruber** [comment]
> Writing from Cloudflare worker directly to InfluxDB may be the simplest option, as we don't need to write any new REST API 💪🏻

It also has the advantage of automatic retention policies

**Comment by pyropy** [comment]
Do we already have InfluxDB deployed? If so I'd appreciate access to the instance 🙏🏻

**Comment by juliangruber** [comment]
Invite sent 👌

**Comment by pyropy** [comment]
I have deployed a cloudflare worker that should report data to influxdb. I have yet to configure it properly now.

 I am proposing is adding another CNAME record (i.e. `stats-test`) on filspark.com domain that will point to `spark-stats` API and setting up cloudflare worker to that route.

After we make sure that cloudflare worker is working correctly, we can setup another CNAME record (i.e. `stats-api`) on filspark.com domain and point it to the `spark-stats` API while configuring cloudflare worker route to original CNAME (`stats`) so all the traffic is routed trough the worker.

**Comment by juliangruber** [comment]
> I am proposing is adding another CNAME record (i.e. stats-test) on filspark.com domain that will point to spark-stats API and setting up cloudflare worker to that route.

This would be a kind of staging environment, right? We can ourselves make requests against that endpoint, and verify they end up in InfluxDB correctly.

As long as the Cloudflare worker doesn't negatively affect requests, I don't see an immediate problem with just using it in prod. Can we just make all requests to api.filspark.com go through your worker?

**Comment by pyropy** [comment]
Yeah, it's kind of a separate environment but In this case I didn't use cloudflare worker environments. Generally it shouldn't affect request performance as metrics are reported after the request has been executed. 

I have created a [repository](https://github.com/pyropy/cf-metrics) which itself is a fork of some other repository I've used as started template. In the end I didn't end up using much of the code inside that repository. I am wondering should we fork my repository as an organization or could I create a new repository and push the code. Given that it's a MIT license I guess it isn't legally binding even if we copy code into new repo?

**Comment by pyropy** [comment]
@juliangruber I've tried adding new CNAME records but it seems like there's some issue with TSL certificates (I get error 525 when trying to send a request). Are the certificates managed by cloudflare or it's done outside of cloudflare?

**Comment by juliangruber** [comment]
> Generally it shouldn't affect request performance as metrics are reported after the request has been executed.

Ok, let's just attach the worker to the production route then (with manual monitoring that nothing goes wrong), for simplicity.

> I have created a [repository](https://github.com/pyropy/cf-metrics) which itself is a fork of some other repository I've used as started template. In the end I didn't end up using much of the code inside that repository. I am wondering should we fork my repository as an organization or could I create a new repository and push the code. Given that it's a MIT license I guess it isn't legally binding even if we copy code into new repo?

Are you able to create new repos in the Filecoin-station org? I think we can then transfer your repository into it. Yes we could just copy the code over, but forks are convenient because you see where the code came from, and that makes it easier to pull in upstream changes.

**Comment by juliangruber** [comment]
> @juliangruber I've tried adding new CNAME records but it seems like there's some issue with TSL certificates (I get error 525 when trying to send a request). Are the certificates managed by cloudflare or it's done outside of cloudflare?

Can we use the existing domain name instead?

**Comment by pyropy** [comment]
> > @juliangruber I've tried adding new CNAME records but it seems like there's some issue with TSL certificates (I get error 525 when trying to send a request). Are the certificates managed by cloudflare or it's done outside of cloudflare?
> 
> Can we use the existing domain name instead?

Yes we could do that, I'm only worried if we're going to miss cache. I've switched worker for few minutes to `stats.filspark.com` this morning and it reported cache status of `DYNAMIC`.

<img width="719" alt="Screenshot 2024-12-10 at 13 47 47" src="https://github.com/user-attachments/assets/a3e18477-4198-4c4e-a666-1970b41d239b">

**Comment by juliangruber** [comment]
> I'm only worried if we're going to miss cache

Does this mean the API cache will be inactive with workers, or that it will be invalidated for every worker update?

**Comment by pyropy** [comment]
> > I'm only worried if we're going to miss cache
> 
> Does this mean the API cache will be inactive with workers, or that it will be invalidated for every worker update?

If the `spark-stats` server remains proxied behind cloudflare (like it currently is) it should still be proxied (hence that's why I wanted to add new CNAME record). Other option would be to cache responses directly in worker and have custom cache behavior.

**Comment by juliangruber** [comment]
I don't understand. We want spark-stats to stay proxied, so that we can use Cloudflare's caching behavior. We want that proxy to also submit usage data to the API, through a Cloudflare worker. What am I missing?

**Comment by pyropy** [comment]
Missing thing is that cloudflare worker and proxy cannot reside at the same route. What I'm suggesting is that we add new route `stats-api.filspark.com` and point it to `spark-stats` server and have cloudflare worker take old route `stats.filspark.com`

![proxy](https://github.com/user-attachments/assets/a23fcb59-e1f3-4565-b678-6fbd7a2f4d5d)

**Comment by juliangruber** [comment]
Got it, I wasn't aware that the worker needs to sit in front of the proxy. This makes sense then. Your graphic helped clarify this 👏 I will take a look at the certificate setup and get back to you

**Comment by juliangruber** [comment]
The problem is that the fly.io deployment requires the hostname to be `stats.filspark.com`. I will look into it there

**Comment by juliangruber** [comment]
`stats-api` isn't the most descriptive name to me, what do you think about `stats-backend`?

Thinking about this more, are you sure this is the right way to go about things? Ie is it a documented pattern to put the worker in front of the proxy? Or are there alternatives, like proxying from inside the worker?

**Comment by juliangruber** [comment]
FYI, https://stats-api.filspark.com works now

**Comment by pyropy** [comment]
> `stats-api` isn't the most descriptive name to me, what do you think about `stats-backend`?
> 
> Thinking about this more, are you sure this is the right way to go about things? Ie is it a documented pattern to put the worker in front of the proxy? Or are there alternatives, like proxying from inside the worker?

We could proxy inside the worker, sure. I guess it would also be more appropriate and also would require a lot less DNS resolutions. I'll see how to use basic cloudflare cache setup with workers.

**Comment by pyropy** [comment]
@juliangruber I have added caching inside cloudflare worker [here](https://github.com/filecoin-station/cf-metrics/pull/1). With that I guess there's no need for additional CNAME records so I'm going to delete them. I've yet to add tests and alter the docs, but the it has been deployed to cloudflare and it seems to be working.

**Comment by bajtos** [comment]
~~Implementing a proxy inside a worker feels rather suboptimal to me. Doesn't Cloudflare CDN offer any hooks we can observe from a worker? In the future, we will want to validate the access token and reject requests with missing or invalid access tokens.~~

_(EDITED) ^^^ That's no longer relevant based on what I learned while writing this comment. ^^^_

Here is some documentation that may be relevant:

Auth with headers:
- https://developers.cloudflare.com/waf/custom-rules/use-cases/configure-token-authentication/
- https://developers.cloudflare.com/workers/examples/auth-with-headers/

Using Cloudflare cache from workers:
- https://developers.cloudflare.com/cache/interaction-cloudflare-products/workers/
- https://developers.cloudflare.com/workers/examples/cache-using-fetch/

Quoting from https://developers.cloudflare.com/workers/reference/how-the-cache-works/
> Conceptually, there are two ways to interact with Cloudflare’s Cache using a Worker:
Call to [fetch()](https://developers.cloudflare.com/workers/runtime-apis/fetch/) in a Workers script. Requests proxied through Cloudflare are cached even without Workers according to a zone’s default or configured behavior (for example, static assets like files ending in .jpg are cached by default). Workers can further customize this behavior by:
> - Setting Cloudflare cache rules (that is, operating on the cf object of a [request](https://developers.cloudflare.com/workers/runtime-apis/request/)).

If I am understanding this topic correctly, then we should have the following architecture:

```
[API client] ---> [Cloudflare worker calling fetch()] --> [Fly.io app]
```

In other words:
- The API client fetches `https://stats.filspark.com/{endpoint}`
- Our Cloudflare Worker is invoked and it fetches `https://spark-stats.fly.dev/{endpoint}`
- Behind the scene, this second `fetch` may be served from Cloudflare's cache (that's how I understand Cloudflare's architecture)

**Comment by pyropy** [comment]
@bajtos I think you might be right about this. To be honest I was confused by this 👇🏻 

> Call to [fetch()](https://developers.cloudflare.com/workers/runtime-apis/fetch/) in a Workers script. Requests proxied through Cloudflare are cached even without Workers according to a zone’s default or configured behavior (for example, static assets like files ending in .jpg are cached by default). Workers can further customize this behavior by...

As I understood it, request is cached if server is proxied behind cloudflare, hence me asking for another CNAME that we could put out server behind.

**Comment by pyropy** [comment]
I have added token to data source on https://spacemeridian.grafana.net/ that reads spark-stats API. Are the dashboards built there are using that data source  also published on https://grafana.filstation.app/?

If that's the case, if I'd to edit API token directly on the dashboard I am afraid we would risk leaking API token.

**Comment by juliangruber** [comment]
> Are the dashboards built there are using that data source also published on https://grafana.filstation.app/?

I don't understand this sentence

**Comment by pyropy** [comment]
> > Are the dashboards built there are using that data source also published on https://grafana.filstation.app/?
> 
> I don't understand this sentence

Woah, I made a mistake. 🤦

I wanted to ask if dashboards built on https://spacemeridian.grafana.net/ are published on https://grafana.filstation.app/ and use data sources defined in https://spacemeridian.grafana.net/?

**Comment by bajtos** [comment]
> I wanted to ask if dashboards built on https://spacemeridian.grafana.net/ are published on https://grafana.filstation.app/ and use data sources defined in https://spacemeridian.grafana.net/?

No, these two Grafana instances are completely independent. If we want to use the same datasource from both instances, then we need to configure it twice.

I believe we don't need to expose API usage data publicly yet (if ever), so there is no need to touch https://grafana.filstation.app/.

**Comment by pyropy** [comment]
All requests to `spark-stats` should be proxied via cloudflare worker and collected data is visualized on grafana dashboard. Since we're not going to setup API key on https://grafana.filstation.app/ I'm going to close this issue.
